crates/agglayer-contracts/src/rollup.rs

```
@@ -0,0 +1,174 @@
+
+                    // Check whether the block number containing the event is now finalized.
+                    if finalized_block_number >= event_block_number {
+                        // Verify that the hash of the block containing
+                        // the event did not change due to potential reorg
+                        let retrieved_block_hash = self
+                            .rpc
+                            .get_block(BlockId::Number(event_block_number.into()))
+                            .await
+                            .ok()
+                            .flatten()
+                            .and_then(|block| block.hash)
+                            .ok_or(L1RpcError::BlockHashNotFound(event_block_number.as_u64()))?;
+
+                        if retrieved_block_hash != event_block_hash {
+                            error!(
+                                "Reorg detected! Retrieved block hash ({:?}) does not match \
+                                 expected event block hash ({:?}).",
+                                retrieved_block_hash, event_block_hash
+                            );
+                            return Err(L1RpcError::ReorgDetected(event_block_number.as_u64()));
+                        }
+
+                        break;
+                    }
+                }
+
+                Ok(())
+            })
+            .await
+            .map_err(|_| {
+                error!(
+                    "Timeout occurred while waiting for block {} to be finalized. Latest \
+                     finalized block: {}",
+                    event_block_number, finalized_block_number
+                );
+                L1RpcError::FinalizationTimeoutExceeded(event_block_number.as_u64())
+            })??;
+        }
+
+        Ok(l1_info_root)
+    }
+
+    async fn get_trusted_sequencer_address(
+        &self,
+        rollup_id: u32,
+        proof_signers: HashMap<u32, Address>,
+    ) -> Result<Address, ()> {
```

Why are we filtering errors here? We should return the proper error.